metric_learn/base_metric.py

```diff
@@ -1,31 +1,76 @@
 from numpy.linalg import cholesky
 from sklearn.base import BaseEstimator
-from sklearn.utils.validation import check_array
+from sklearn.utils.validation import check_array, _is_arraylike
 from sklearn.metrics import roc_auc_score
 import numpy as np
 from abc import ABCMeta, abstractmethod
 import six
-from ._util import check_tuples
+from ._util import check_tuples, SimplePreprocessor
 
 
 class BaseMetricLearner(six.with_metaclass(ABCMeta, BaseEstimator)):
 
-  @abstractmethod
-  def score_pairs(self, pairs):
-    """Returns the score between pairs
-    (can be a similarity, or a distance/metric depending on the algorithm)
+  def __init__(self, preprocessor=None):
+    """
 
     Parameters
     ----------
-    pairs : `numpy.ndarray`, shape=(n_samples, 2, n_features)
-      3D array of pairs.
+    preprocessor : array-like, shape=(n_samples, n_features) or callable
+        The preprocessor to call to get tuples from indices. If array-like,
+        tuples will be gotten like this: X[indices].
+    """
+    self.preprocessor = preprocessor
```

Comment:
couldn't the `self.check_preprocessor()` be called here (or in the relevant meta classes) once and for all? Instead of calling it in various places in the downstream classes?

Reply:
Yes I agree it is very redundant
Maybe we should create an abstract `fit` for a class `WeaklySupervisedMixin` and put it here, because to be compatible with scikit-learn `set_params` used in grid-searches etc, I think the `__init__` should not contain any processing of the input (like `if preprocessor [...]`  `else` etc), and they should all be done in `fit`. (Even if this was in order to use `set_params` to do hyper-parameter search in grid-search, and we shouldn't do grid search on a preprocessor too much (except maybe if we use the preprocessor as a feature extractor and we want to test different feature extractors))

Reply:
After thinking about it I am not sure what to do because checking the preprocessor in an abstract `_fit` in say `BaseMetricLearner` would mean calling `super()._fit` in `_BaseMMC._fit` and in `NCA.fit` for instance (btw the class call and super call would have different names for NCA: `fit` and `_fit`, respectively, but maybe this is not a problem), which would amount to the same lines of code added as before. But maybe this is a better practice and would pave the way if we have to mutualize other stuff in `fit` ? Or maybe for now it adds a little complexity with super calls, instead of an explicit `check_preprocessor` ? I'm OK with both strategies

Reply:
In fact the right place for the call to `check_preprocessor` should probably be in a common `_prepare_inputs` function (see more detailed comment below).